aten/src/ATen/native/cuda/SpectralOps.cu

```diff
@@ -0,0 +1,274 @@
+  for (int64_t i = signal_ndim - 1; i >= 0; i--) {
+    signal_size = checked_signal_sizes[i];
+    istride = input.stride(i + 1);
+    if (is_half && !is_pow_of_two(signal_size)) {
+      std::ostringstream ss;
+      ss << "cuFFT doesn't support signals of half type with size that is not "
+         << "a power of two, but gets a signal size of " << signal_size << " at"
+         << "signal dimension " << i;
+      throw std::runtime_error(ss.str());
+    }
+    signal_sizes[i] = signal_size;
+    if (!need_contiguous) {
+      // set the inembeds for dim in last dimension
+      if (ilast_stride == 0) {
+        need_contiguous = istride != 0;
+        if (i < signal_ndim - 1) {
+          inembeds[i + 1] = 0;
+        }
+      } else {
+        need_contiguous = istride % ilast_stride != 0;
+        if (i < signal_ndim - 1) {
+          inembeds[i + 1] = istride / ilast_stride;
+        }
+      }
+      ilast_stride = istride;
+    }
+  }
+
+  need_contiguous |= (input.size(0) > 1 && input.stride(0) == 0);
+  if (need_contiguous) {
+    input = input.contiguous();
+    auto input_sizes = input.sizes();
+    std::copy(&input_sizes[1], &input_sizes[signal_ndim] + 1, &inembeds[0]);
+  }
+
+  // set base_istride (stride at innermost dim of signal)
+  // set idist (stride at batch dim)
+  long long int idist = complex_input ? input.stride(0) >> 1 : input.stride(0);
+  // Even if batch dimension is one, then idist (stride(0)) doesn't matter,
+  // cuFFT errors if idist = 0. This is hack to make it succeed.
+  if (idist == 0 && batch == 1) {
+    idist = 1;
+  }
+  long long int base_istride = complex_input ? input.stride(signal_ndim) >> 1 : input.stride(signal_ndim);
+
+  // output, base_ostride, odist
+  auto output = input.type().tensor(output_sizes);
+  std::vector<long long int> onembeds(&output_sizes[1], &output_sizes[signal_ndim] + 1);
+  long long int odist = complex_output ? output.stride(0) >> 1 : output.stride(0);
+  long long int base_ostride = 1;
+
+  CufftHandle plan;
+  size_t ws = 0;
+  cudaDataType itype, otype, exec_type;
+  if (input.type().scalarType() == ScalarType::Float) {
+    itype = complex_input ? CUDA_C_32F : CUDA_R_32F;
+    otype = complex_output ? CUDA_C_32F : CUDA_R_32F;
+    exec_type = CUDA_C_32F;
+  } else if (input.type().scalarType() == ScalarType::Double) {
+    itype = complex_input ? CUDA_C_64F : CUDA_R_64F;
+    otype = complex_output ? CUDA_C_64F : CUDA_R_64F;
+    exec_type = CUDA_C_64F;
+  } else if (input.type().scalarType() == ScalarType::Half) {
+    itype = complex_input ? CUDA_C_16F : CUDA_R_16F;
+    otype = complex_output ? CUDA_C_16F : CUDA_R_16F;
+    exec_type = CUDA_C_16F;
+  } else {
+    std::ostringstream ss;
+    ss << "cuFFT doesn't support tensor of type: "
+       << at::toString(input.type().scalarType());
+    throw std::runtime_error(ss.str());
+  }
+
+  CUFFT_CHECK(cufftXtMakePlanMany(plan.get(), signal_ndim, &signal_sizes[0], &inembeds[0],
```

Comment:
We're C++11, you can use `signal_sizes.data()`

Reply:
It's worth noting that plan creation is not free. It may be profitable to cache plans; hard to say without measurements about how expensive making these plans are.